frame/nomination-pools/src/lib.rs

```diff
@@ -1848,41 +1850,37 @@ pub mod pallet {
 		pub fn update_roles(
 			origin: OriginFor<T>,
 			pool_id: PoolId,
-			root: ConfigOp<T::AccountId>,
-			nominator: ConfigOp<T::AccountId>,
-			state_toggler: ConfigOp<T::AccountId>,
+			root: Option<T::AccountId>,
+			nominator: Option<T::AccountId>,
+			state_toggler: Option<T::AccountId>,
 		) -> DispatchResult {
 			let o1 = origin;
 			let o2 = o1.clone();
 
-			let is_pool_root = || -> Result<BondedPool<T>, sp_runtime::DispatchError> {
+			let mut bonded_pool = BondedPool::<T>::get(pool_id).ok_or(Error::<T>::PoolNotFound)?;
+			let is_pool_root = || -> Result<(), sp_runtime::DispatchError> {
 				let who = ensure_signed(o1)?;
-				let bonded_pool = BondedPool::<T>::get(pool_id).ok_or(Error::<T>::PoolNotFound)?;
-				ensure!(bonded_pool.roles.root == who, Error::<T>::DoesNotHavePermission);
-				Ok(bonded_pool)
+				ensure!(bonded_pool.can_update_roles(&who), Error::<T>::DoesNotHavePermission);
+				Ok(())
 			};
-			let is_root = || -> Result<BondedPool<T>, sp_runtime::DispatchError> {
+			let is_root = || -> Result<(), sp_runtime::DispatchError> {
 				ensure_root(o2)?;
-				let bonded_pool = BondedPool::<T>::get(pool_id).ok_or(Error::<T>::PoolNotFound)?;
-				Ok(bonded_pool)
+				Ok(())
 			};
 
-			let mut bonded_pool = is_root().or_else(|_| is_pool_root())?;
```

Comment:
I don't see an obvious way to do this better.  
Maybe with some `match (ensure_signed, ensure_root)` or something but it would also be complicated.